test/distributed/test_ddp_with_rpc.py

```diff
@@ -0,0 +1,283 @@
+            world_size=cls.NUM_TRAINERS + 1,
+            group_name="trainers/remote RPC",
+        ):
+            pass
+        gLogger.info(f"Exiting remote worker.")
+
+    def spawn_remote_worker(self):
+        remote_worker_process = multiprocessing.Process(
+            target=self._remote_worker, name=self.REMOTE_WORKER_NAME
+        )
+        remote_worker_process.start()
+        self.processes.append(remote_worker_process)
+
+    def setUp(self):
+        super(TestDdpWithRpc, self).setUp()
+
+        os.environ["MASTER_ADDR"] = str(MASTER_ADDR)
+        os.environ["MASTER_PORT"] = str(MASTER_PORT)
+        # os.environ["WORLD_SIZE"] = str(self.NUM_TRAINERS + 1)
+
+        self.processes: List[multiprocessing.Process] = []
+        self.remote_em_rref: rpc.RRef = None
+
+        self.spawn_remote_worker()
+
+    def tearDown(self):
+        super(TestDdpWithRpc, self).tearDown()
+        self.join_processes()
+        for p in self.processes:
+            p.terminate()
+
+    def join_processes(self):
+        for p in self.processes:
+            p.join()
+
+    # TODO: make it work
+    def run_one_mini_batch(self):
+        with dist_autograd.context() as context_id:
+            # Forward pass (create references on remote nodes).
+            rref1 = rpc.remote(dst_name, random_tensor)
+            rref2 = rpc.remote(dst_name, random_tensor)
+            loss = rref1.to_here() + rref2.to_here()
+
+            # Backward pass (run distributed autograd).
+            dist_autograd.backward([loss.sum()])
+
+            # Build DistributedOptimizer.
+            dist_optim = DistributedOptimizer(
+                optim.SGD, [rref1, rref2], lr=0.05
+            )
+
+            # Run the distributed optimizer step.
+            dist_optim.step()
+
+    @classmethod
+    def _trainer(
+        cls,
+        remote_em_rref: rpc.RRef,
+        rank: int,
+        func: Callable[[nn.Module], None],
+        ddp_mode: DdpMode,
+    ):
+        trainer_name = cls.TRAINER_NAME_TEMPLATE.format(rank)
+        gLogger.info(f"Starting trainer {trainer_name}")
+        if ddp_mode in (DdpMode.OUTSIDE, DdpMode.INSIDE):
+            gLogger.info(f"Initing trainer process group for DDP.")
+            dist.init_process_group(
+                # TODO: test other rpc backend.
+                "gloo",
+                group_name=cls.TRAINER_GROUP,
+                rank=rank,
+                world_size=cls.NUM_TRAINERS,
+            )
+
+        rpc_context = RpcContext(
+            name=trainer_name,
+            rank=rank,
+            world_size=cls.NUM_TRAINERS + 1,
+            group_name="trainers/remote RPC",
+        )
+        if remote_em_rref is None:
+            # TODO: All trainers should share the same remote embedding table.
+            gLogger.warning(f"Creating a remote em for trainer #{rank}")
+            remote_em_rref = rpc.remote(
+                cls.REMOTE_WORKER_NAME, RemoteEM, args=(NUM_EM_ROW, D_SPARSE)
```

Comment:
> I saw the TODO above. In the current version, every trainer has its dedicated EM, right?

Correct.

> I agree that it will make it easier if we can decouple RRef with RPC. Before we having that feature ready, would it be acceptable that we let a master driving all the work? For example, all trainers and the ps can just block on `rpc.shutdown()`, and then the master first creates the EM rref and then share it will all trainers.

That's doable. Any example code for sharing a rref between 2 processes? (maybe pickle?)